lms/views/admin.py

```diff
@@ -0,0 +1,91 @@
+    def __init__(self, request):
+        self.request = request
+        self.application_instance = request.find_service(name="application_instance")
+
+    @view_config(
+        route_name="admin.index",
+        renderer="lms:templates/admin/index.html.jinja2",
+        permission=Permissions.ADMIN,
+    )  # pylint: disable=no-self-use
+    def index(self):
+        return {}
+
+    @view_config(
+        route_name="admin.installations",
+        renderer="lms:templates/admin/installations.html.jinja2",
+        permission=Permissions.ADMIN,
+    )  # pylint: disable=no-self-use
+    def installations(self):
+        return {}
+
+    @view_config(
+        route_name="admin.installations",
+        permission=Permissions.ADMIN,
+        request_method="POST",
+    )
+    def find_installation(self):
+        if "query" not in self.request.params:
+            self.request.session.flash("Missing mandatory 'query'", "errors")
+            return HTTPFound(location=self.request.route_url("admin.installations"))
+
+        installation = self.application_instance.find(self.request.params["query"])
+        if installation:
+            return HTTPFound(
+                location=self.request.route_url(
+                    "admin.installation", id=installation.id
+                ),
+            )
+        self.request.session.flash(
+            f'No installation found for {self.request.params["query"]}', "errors"
+        )
+        return HTTPFound(location=self.request.route_url("admin.installations"))
+
+    @view_config(
+        route_name="admin.installation",
+        renderer="lms:templates/admin/installation.html.jinja2",
+        permission=Permissions.ADMIN,
+    )
+    def show_installation(self):
+
+        installation = self.application_instance.get(self.request.matchdict["id"])
+        return {"installation": installation}
+
+    @view_config(
+        route_name="admin.installation",
+        permission=Permissions.ADMIN,
+        request_method="POST",
+    )
+    def update_installation(self):
+        installation = self.application_instance.get(self.request.matchdict["id"])
+
+        sections_enabled = (
+            "sections_enabled" in self.request.params
+            and self.request.params["sections_enabled"] == "on"
+        )
+        groups_enabled = (
+            "groups_enabled" in self.request.params
+            and self.request.params["groups_enabled"] == "on"
+        )
+
+        self.application_instance.update_settings(
+            installation,
+            canvas_sections_enabled=sections_enabled,
+            canvas_groups_enabled=groups_enabled,
+        )
+        self.request.session.flash(
+            f"Updated installation {installation.id}", "messages"
+        )
+
+        return HTTPFound(location=self.request.route_url("admin.installations"))
+
+    @forbidden_view_config()
+    def logged_out(self):
+        return HTTPFound(location=self.request.route_url("pyramid_googleauth.login"))
```

Comment:
I'm not sure this forbidden view is doing what you might think. Since no arguments are passed to `@forbidden_view_config()` (except the `request_method="GET"` from the `@view_defaults`) this forbidden view is being registered for _all_ `GET` requests that fail a permissions check, for any route. So this will cause all permissions failures to redirect to the admin login page.

The forbidden view is also _not_ being registered for the `update_installion()` form submission because it's a `POST`.

I think you need a forbidden view that's registered for the admin pages only. You need to find the right view config predicates: https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html#predicate-arguments The easiest option is probably `@forbidden_view_config(path_info="/admin/*")`. I know people don't like to reference parts of the URL path in the code but I think it's the only way to get the correct behaviour.

I think the forbidden view should be moved out of the `AdminViews` class as well, which will allow the `permission=Permissions.ADMIN` to be moved into the `@view_defaults`

Reply:
Yep that was completely broken, `path_info` seems the most straightforward fix :+1: